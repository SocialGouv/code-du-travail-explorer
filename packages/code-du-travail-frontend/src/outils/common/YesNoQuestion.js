import React from "react";
import PropTypes from "prop-types";
import { Field } from "react-final-form";
import { OnChange } from "react-final-form-listeners";
import { InputRadio } from "@socialgouv/react-ui";
import { RadioContainer } from "./stepStyles";
import { Question } from "./Question";
import { requiredBoolean } from "./validators";
import { ErrorField } from "./ErrorField";

function YesNoQuestion({
  name,
  label,
  required = true,
  onChange,
  ...otherProps
}) {
  return (
    <>
      <Question as="p" required={required}>
        {label}
      </Question>
      <RadioContainer {...otherProps}>
        <Field
          type="radio"
          parse={value => value === "true"}
          name={name}
          label="Oui"
          value={true}
          validate={requiredBoolean}
        >
          {props => (
            <InputRadio id={`${name}-yes`} label="Oui" {...props.input} />
          )}
        </Field>
        <Field
          type="radio"
          parse={value => value === "true"}
          name={name}
          value={false}
          validate={requiredBoolean}
        >
          {props => (
            <InputRadio id={`${name}-no`} label="Non" {...props.input} />
          )}
        </Field>
      </RadioContainer>
      <ErrorField name={name} />
      {onChange && (
        <OnChange name={name}>{values => onChange(values)}</OnChange>
      )}
    </>
  );
}
YesNoQuestion.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired
};

export { YesNoQuestion };
