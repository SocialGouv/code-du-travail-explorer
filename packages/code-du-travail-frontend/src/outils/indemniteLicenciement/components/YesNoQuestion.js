import React from "react";
import PropTypes from "prop-types";
import { Field } from "react-final-form";
import { OnChange } from "react-final-form-listeners";
import { Alert } from "@cdt/ui";
import { Label, RadioContainer, QuestionParagraphe } from "../stepStyles";
import { requiredBoolean } from "../validators";

function YesNoQuestion({ name, label, onChange }) {
  return (
    <>
      <QuestionParagraphe>{label}</QuestionParagraphe>
      <RadioContainer>
        <Label>
          <Field
            parse={value => value === "true"}
            component="input"
            type="radio"
            name={name}
            value={true}
            validate={requiredBoolean}
          />
          <span>Oui</span>
        </Label>
        <Label>
          <Field
            parse={value => value === "true"}
            component="input"
            type="radio"
            name={name}
            value={false}
            validate={requiredBoolean}
          />
          <span>Non</span>
        </Label>
      </RadioContainer>
      <Field
        name={name}
        subscribe={{ error: true, visited: true, touched: true }}
        render={({ meta: { visited, touched, error } }) =>
          (error && visited) || (error && touched) ? (
            <Alert>{error}</Alert>
          ) : null
        }
      />
      {onChange && (
        <OnChange name={name}>{values => onChange(values)}</OnChange>
      )}
    </>
  );
}
YesNoQuestion.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired
};

export { YesNoQuestion };
