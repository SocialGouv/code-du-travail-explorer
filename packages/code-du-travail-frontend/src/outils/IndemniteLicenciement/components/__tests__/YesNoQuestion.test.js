import React from "react";
import { render, fireEvent } from "@wrapped-testing-library/react";
import { YesNoQuestion } from "../YesNoQuestion";
import { Form } from "react-final-form";

describe("<YesNoQuestion />", () => {
  it("should render", () => {
    const onSubmit = jest.fn();
    const { container } = render(
      <Form
        onSubmit={onSubmit}
        render={() => <YesNoQuestion name="test" label="lorem ipsum ?" />}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should call onChange", () => {
    const onSubmit = jest.fn();
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <Form
        onSubmit={onSubmit}
        render={() => (
          <>
            <YesNoQuestion
              name="test"
              label="lorem ipsum ?"
              onChange={onChange}
            />
          </>
        )}
      />
    );
    const oui = getByLabelText(/oui/i);
    oui.click();
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it("should render error when form validate", () => {
    const onSubmit = jest.fn();
    const { getByText } = render(
      <Form onSubmit={onSubmit}>
        {({ handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <YesNoQuestion name="test" label="lorem ipsum ?" />
            <button>validate</button>
          </form>
        )}
      </Form>
    );
    const non = getByText(/validate/i);
    fireEvent.click(non);
    expect(getByText(/ce champ est requis/i)).toBeDefined();
  });

  it("should render error when radio change", () => {
    const onSubmit = jest.fn();
    const { getByLabelText, getByText } = render(
      <Form
        onSubmit={onSubmit}
        validate={values =>
          values.test === "oui" ? {} : { test: "mauvaise réponse" }
        }
      >
        {({ handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <YesNoQuestion name="test" label="lorem ipsum ?" />
            <button>validate</button>
          </form>
        )}
      </Form>
    );
    const non = getByLabelText(/non/i);
    fireEvent.click(non);
    expect(getByText(/mauvaise réponse/i)).toBeDefined();
  });
});
