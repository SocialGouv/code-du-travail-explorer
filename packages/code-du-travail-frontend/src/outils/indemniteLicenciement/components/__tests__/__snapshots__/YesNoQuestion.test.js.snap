// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YesNoQuestion /> should render 1`] = `
.c2 {
  padding: 0;
  text-align: center;
  font-size: 1.24rem;
  margin-right: 2em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  font-size: 1.1rem;
  display: block;
  margin-top: 1.25rem;
  margin-bottom: 0.625rem;
}

<div>
  <p
    class="c0"
  >
    lorem ipsum ?
  </p>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        name="test"
        type="radio"
        value="true"
      />
      <span>
        Oui
      </span>
    </label>
    <label
      class="c2"
    >
      <input
        name="test"
        type="radio"
        value="false"
      />
      <span>
        Non
      </span>
    </label>
  </div>
</div>
`;
